### ISSUE TITLE
GDNative should provide a way to secure void* pointer casts

### ISSUE THREAD

**Issue by Zylann** [godotengine/godot#15918]
Godot 3.0

The NativeScript API exposes `godot_nativescript_get_userdata`, which gives a `void*`. Like all naked pointers, it cannot be validated before being casted, especially in this API because many GDNative libraries could make use of it at the same time.

Currently, the C++ bindings cast this `void*` into the custom corresponding GodotScript class, without verifying wether it's valid or not, leading to high crash potential for users.

There may be a safer way to do this, which is to add typetags along these pointers, where the said typetags would map to a unique memory address in the libraries (or generated by GDNative), as I explained in the following issue https://github.com/GodotNativeTools/godot-cpp/issues/77

**Comment by karroffel** [comment]
The D bindings check if a method with a certain ID as the name is defined, then it knows it's the proper type!

In C++ we could do something similar using get_meta() maybe?

**Comment by karroffel** [comment]
Because NativeScript is only concerned with `Object`s anyway I think using `get_meta()` should be good enough

**Comment by sheepandshepherd** [comment]
Most casts in Godot-D are type-checked, yeah. I used a custom method for D classes, and `is_class` (like Zylann mentioned) for Godot classes.

But that's a lot of overhead just for type safety, so it would be nice to have a more efficient way eventually (in 3.1?). This is how the engine checks in `cast_to`:
```cpp
		if (p_object->is_class_ptr(T::get_class_ptr_static()))
			return static_cast<T *>(p_object);
```
(`is_class_ptr` also checks parent types.) If we could retrieve [the unique static pointer](https://github.com/godotengine/godot/blob/ef1c26a9a5882e7d1b8027958eec6fcc82a46c44/core/object.h#L258) from each class, we could do the same in GDNative. The difficulty is just in how to expose this to GDNative.

**Comment by karroffel** [comment]
implemented by #17980 and #17980
